explorer/interpreter/type_checker.cpp

```diff
@@ -5071,6 +5070,29 @@ auto TypeChecker::DeclareDeclaration(Nonnull<Declaration*> d,
       auto& var = cast<VariableDeclaration>(*d);
       // Associate the variable name with it's declared type in the
       // compile-time symbol table.
+
+
+      if (llvm::isa<AutoPattern>(var.binding().type())) {
```

Comment:
Right now we have handling for `var` typing in two places:

TypeCheckStmt for VariableDefinition:
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/type_checker.cpp#L3753

TypeCheckDeclaration for VariableDeclaration (incomplete):
https://github.com/carbon-language/carbon-lang/blob/trunk/explorer/interpreter/type_checker.cpp#L4973

Note the "// TODO: consider adding support for `auto`" on 4980 in particular.

I think the right thing to do here is to try sharing code with the VariableDefinition. Note too, the purpose of the optional in TypeCheckPattern's signature is to allow the type to be checked -- that conditional support is what's being used on line 3759 for optional.

I'm not sure if the type-checking code can easily be shifted to TypeCheckDeclaration (which, long-term, it definitely should be). However, it's probably best to share a little more of the type checking implementation with VariableDefinition: the overall flow and behavior should match, so a helper function may help avoid a split in implementation behaviors (which I think is at the root of the issue affecting global variables).